### PR TITLE
Fix #20577: Add windows symbols

### PR DESCRIPTION
**Pull request by etcimon**
Merge of #2937 #2938 #2939

### CODE REVIEW

**Comment by Geod24** [comment]
https://github.com/dlang/druntime/blob/master/CONTRIBUTING.md

File: src/core/sys/windows/sdkddkver.d [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+else static if (_WIN32_WINNT <= _WIN32_WINNT_WS03)
+   enum _WIN32_IE    =  _WIN32_IE_WS03;
+else static if (_WIN32_WINNT <= _WIN32_WINNT_VISTA)
+   enum _WIN32_IE    =  _WIN32_IE_LONGHORN;
+else static if (_WIN32_WINNT <= _WIN32_WINNT_WIN7)
+   enum _WIN32_IE    =  _WIN32_IE_WIN7;
+else static if (_WIN32_WINNT <= _WIN32_WINNT_WIN8)
+   enum _WIN32_IE    =  _WIN32_IE_WIN8;
+else
+   enum _WIN32_IE       = 0x0A00;
```
Comment (CyberShadow) [inline src/core/sys/windows/sdkddkver.d]: I think we already have this here:

https://github.com/dlang/druntime/blob/d7800b516fd31eb980fb64ce99d02f22c75820d9/src/core/sys/windows/w32api.d#L48-L84

File: src/core/sys/windows/winbase.d [hunk 2]
```diff
@@ -2257,7 +2257,8 @@ WINBASEAPI BOOL WINAPI SetEvent(HANDLE);
     BOOL IsTextUnicode(PCVOID, int, LPINT);
     BOOL IsValidAcl(PACL);
     BOOL IsValidSecurityDescriptor(PSECURITY_DESCRIPTOR);
-    BOOL IsValidSid(PSID);
+    BOOL IsValidSid(PSID);    
+   BOOL CreateWellKnownSid(WELL_KNOWN_SID_TYPE, PSID, PSID, PDWORD);
```
Comment (CyberShadow) [inline src/core/sys/windows/winbase.d]: Whitespace

File: src/core/sys/windows/winnt.d [hunk 3]
```diff
@@ -2436,6 +2437,10 @@ struct TOKEN_USER {
 }
 alias TOKEN_USER* PTOKEN_USER;
 
+struct TOKEN_MANDATORY_LABEL {
+   SID_AND_ATTRIBUTES Label;
```
Comment (CyberShadow) [inline src/core/sys/windows/winnt.d]: Whitespace. We use an indent of 4 spaces in Phobos.

File: src/core/sys/windows/winnt.d [hunk 4]
```diff
@@ -2451,27 +2456,63 @@ struct SECURITY_DESCRIPTOR {
     PACL Dacl;
 }
 alias SECURITY_DESCRIPTOR* PSECURITY_DESCRIPTOR, PISECURITY_DESCRIPTOR;
+enum TOKEN_ELEVATION_TYPE {
+   TokenElevationTypeDefault = 1,
+   TokenElevationTypeFull,
+   TokenElevationTypeLimited
```
Comment (CyberShadow) [inline src/core/sys/windows/winnt.d]: Ditto

File: mak/WINDOWS [hunk 5]
```diff
@@ -1202,6 +1202,9 @@ $(IMPDIR)\core\sys\windows\rpcnterr.d : src\core\sys\windows\rpcnterr.d
 
 $(IMPDIR)\core\sys\windows\schannel.d : src\core\sys\windows\schannel.d
 	copy $** $@
+
```
Comment (CyberShadow) [inline mak/WINDOWS]: Please remove the tab character from this line. Trailing whitespace is not allowed by CI rules.

File: mak/WINDOWS [hunk 6]
```diff
@@ -1203,6 +1203,9 @@ $(IMPDIR)\core\sys\windows\rpcnterr.d : src\core\sys\windows\rpcnterr.d
 $(IMPDIR)\core\sys\windows\schannel.d : src\core\sys\windows\schannel.d
 	copy $** $@
 
+$(IMPDIR)\core\sys\windows\sdkddkver.d : src\core\sys\windows\sdkddkver.d
+	copy $** $@
+
```
Comment (CyberShadow) [inline mak/WINDOWS]: I believe you also need to add the file to `mak/COPY` and `mak/SRCS`. See the contribution guide, or where else existing files occur (e.g. the `secext.d` below).

File: src/core/sys/windows/winbase.d [hunk 7]
```diff
@@ -2257,7 +2257,8 @@ WINBASEAPI BOOL WINAPI SetEvent(HANDLE);
     BOOL IsTextUnicode(PCVOID, int, LPINT);
     BOOL IsValidAcl(PACL);
     BOOL IsValidSecurityDescriptor(PSECURITY_DESCRIPTOR);
-    BOOL IsValidSid(PSID);
+    BOOL IsValidSid(PSID);
```
Comment (CyberShadow) [inline src/core/sys/windows/winbase.d]: Trailing whitespace, here and elsewhere (see CI log).

**Comment by etcimon** [comment]
I don't have a bugzilla number though, any chance I could get this merged anyways?

**Comment by CyberShadow** [comment]
> I don't have a bugzilla number though

Just create an issue. This is one of two ways that changes get publicly documented (which all non-trivial ones should be). The other way is generally left for bigger changes, as they get a better spot in the changelog.

> any chance I could get this merged anyways?

No, but I can finish the rest for you if you like.

**Comment by etcimon** [comment]
> No, but I can finish the rest for you if you like.

To test this, I used the some D translation I made of this code: https://github.com/modzero/fix-windows-privacy/blob/master/fix-privacy-base/SelfElevate.cpp

How would I go making an issue without copying my D version of it?

**Comment by CyberShadow** [comment]
The issue only needs to hold a short description of the goal that this PR aims to achieve. You could add more details such as your motivation for pursuing said goal, or include D code that currently doesn't work but you wish will work, but it's not necessary. The goal is that D users who read the changelog would have something to click and read in order to understand why the change was performed and what the benefit that it brought to D was.

**Comment by etcimon** [comment]
The relevant issue has been added: https://issues.dlang.org/show_bug.cgi?id=20577

**Comment by CyberShadow** [comment]
Great. Now you need to edit the commit message (not PR title/description) to reference the issue number. E.g. you could use:

```
Fix issue 20577 - Add missing symbols related to Windows UAC
```

**Comment by etcimon** [comment]
Should I use the issue number in the commit description as well? Or just the title of the PR is fine?

**Pull request merged**